torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -343,10 +379,33 @@ def _init_param_attributes(self, p: Parameter) -> None:
             p, "_orig_size"
         ), "Parameters should have been sharded during construction."
         if hasattr(p, "_local_shard"):
+            # If CPU offloading, p._local_shard should have been placed on CPU
+            # during its first lazy construction.
+            if self.cpu_offload.offload_params:
+                assert (
+                    p._local_shard.device == torch.device("cpu")  # type: ignore[attr-defined]
+                ), "Expected p._local_shard to be on CPU."  # type: ignore[attr-defined]
             return
 
-        # A single shard of the parameters.
+        # A single shard of the parameters. Also makes p._local_shard to be on
+        # CPU if we are CPU offloading, since p.data would be on CPU during
+        # init.
+        if self.cpu_offload.offload_params:
```

Comment:
would I be correct that the implementation is the same as the following?

```python
if self.cpu_offload.offload_params:
  if hasattr(p, "_local_shard"):
    assert (
      p._local_shard.device == torch.device("cpu")  # type: ignore[attr-defined]
    ), "Expected p._local_shard to be on CPU."  # type: ignore[attr-defined]
    return

  assert p.device == torch.device("cpu"), "Expected param to be on CPU when cpu_offloading is enabled."
  p._local_shard = p.data  # type: ignore[attr-defined]
  p._local_shard.pin_memory()  # type: ignore[attr-defined]
  p._cpu_grad = torch.zeros_like(  # type: ignore[attr-defined]
    ...
  ).pin_memor()
else:
  p._local_shard = p.data  # type: ignore[attr-defined]
  

  
```

Reply:
Yep, although in the above code, the part before and until `return` is redundant right? We check that above after L381? I feel like the current code is a bit more clear that `p._local_shard = p.data` always regardless of CPU offload or not, but open to changing it to this as well depending on what you think.